File: sdk/go/datastore_server.go
package gestalt

import (
	"context"

	proto "github.com/valon-technologies/gestalt/sdk/go/gen/v1"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/emptypb"
)

type datastoreServer struct {
	proto.UnimplementedDatastoreProviderServer
	store DatastoreProvider
}

func newDatastoreProviderServer(store DatastoreProvider) *datastoreServer {
	return &datastoreServer{store: store}
}

func (s *datastoreServer) Migrate(ctx context.Context, _ *emptypb.Empty) (*emptypb.Empty, error) {
	if err := s.store.Migrate(ctx); err != nil {
		return nil, providerRPCError("migrate", err)
	}
	return &emptypb.Empty{}, nil
}

func (s *datastoreServer) GetUser(ctx context.Context, req *proto.GetUserRequest) (*proto.StoredUser, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	user, err := s.store.GetUser(ctx, req.GetId())
	if err != nil {
		return nil, providerRPCError("get user", err)
	}
	if user == nil {
		return nil, status.Error(codes.NotFound, "user not found")
	}
	return user, nil
}

func (s *datastoreServer) FindOrCreateUser(ctx context.Context, req *proto.FindOrCreateUserRequest) (*proto.StoredUser, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	user, err := s.store.FindOrCreateUser(ctx, req.GetEmail())
	if err != nil {
		return nil, providerRPCError("find or create user", err)
	}
	if user == nil {
		return nil, status.Error(codes.Internal, "datastore provider returned nil user")
	}
	return user, nil
}

func (s *datastoreServer) PutStoredIntegrationToken(ctx context.Context, req *proto.StoredIntegrationToken) (*emptypb.Empty, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	if err := s.store.PutIntegrationToken(ctx, req); err != nil {
		return nil, providerRPCError("put integration token", err)
	}
	return &emptypb.Empty{}, nil
}

func (s *datastoreServer) GetStoredIntegrationToken(ctx context.Context, req *proto.GetStoredIntegrationTokenRequest) (*proto.StoredIntegrationToken, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	token, err := s.store.GetIntegrationToken(ctx, req.GetUserId(), req.GetIntegration(), req.GetConnection(), req.GetInstance())
	if err != nil {
		return nil, providerRPCError("get integration token", err)
	}
	if token == nil {
		return nil, status.Error(codes.NotFound, "integration token not found")
	}
	return token, nil
}

func (s *datastoreServer) ListStoredIntegrationTokens(ctx context.Context, req *proto.ListStoredIntegrationTokensRequest) (*proto.ListStoredIntegrationTokensResponse, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	tokens, err := s.store.ListIntegrationTokens(ctx, req.GetUserId(), req.GetIntegration(), req.GetConnection())
	if err != nil {
		return nil, providerRPCError("list integration tokens", err)
	}
	return &proto.ListStoredIntegrationTokensResponse{Tokens: tokens}, nil
}

func (s *datastoreServer) DeleteStoredIntegrationToken(ctx context.Context, req *proto.DeleteStoredIntegrationTokenRequest) (*emptypb.Empty, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	if err := s.store.DeleteIntegrationToken(ctx, req.GetId()); err != nil {
		return nil, providerRPCError("delete integration token", err)
	}
	return &emptypb.Empty{}, nil
}

func (s *datastoreServer) PutAPIToken(ctx context.Context, req *proto.StoredAPIToken) (*emptypb.Empty, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	if err := s.store.PutAPIToken(ctx, req); err != nil {
		return nil, providerRPCError("put api token", err)
	}
	return &emptypb.Empty{}, nil
}

func (s *datastoreServer) GetAPITokenByHash(ctx context.Context, req *proto.GetAPITokenByHashRequest) (*proto.StoredAPIToken, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	token, err := s.store.GetAPITokenByHash(ctx, req.GetHashedToken())
	if err != nil {
		return nil, providerRPCError("get api token by hash", err)
	}
	if token == nil {
		return nil, status.Error(codes.NotFound, "api token not found")
	}
	return token, nil
}

func (s *datastoreServer) ListAPITokens(ctx context.Context, req *proto.ListAPITokensRequest) (*proto.ListAPITokensResponse, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	tokens, err := s.store.ListAPITokens(ctx, req.GetUserId())
	if err != nil {
		return nil, providerRPCError("list api tokens", err)
	}
	return &proto.ListAPITokensResponse{Tokens: tokens}, nil
}

func (s *datastoreServer) RevokeAPIToken(ctx context.Context, req *proto.RevokeAPITokenRequest) (*emptypb.Empty, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	if err := s.store.RevokeAPIToken(ctx, req.GetUserId(), req.GetId()); err != nil {
		return nil, providerRPCError("revoke api token", err)
	}
	return &emptypb.Empty{}, nil
}

func (s *datastoreServer) RevokeAllAPITokens(ctx context.Context, req *proto.RevokeAllAPITokensRequest) (*proto.RevokeAllAPITokensResponse, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	revoked, err := s.store.RevokeAllAPITokens(ctx, req.GetUserId())
	if err != nil {
		return nil, providerRPCError("revoke all api tokens", err)
	}
	return &proto.RevokeAllAPITokensResponse{Revoked: revoked}, nil
}

func (s *datastoreServer) GetOAuthRegistration(ctx context.Context, req *proto.GetOAuthRegistrationRequest) (*proto.OAuthRegistration, error) {
	store, ok := s.store.(OAuthRegistrationStore)
	if !ok {
		return nil, providerRPCError("get oauth registration", ErrOAuthRegistrationStoreUnsupported)
	}
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	registration, err := store.GetOAuthRegistration(ctx, req.GetAuthServerUrl(), req.GetRedirectUri())
	if err != nil {
		return nil, providerRPCError("get oauth registration", err)
	}
	if registration == nil {
		return nil, status.Error(codes.NotFound, "oauth registration not found")
	}
	return registration, nil
}

func (s *datastoreServer) PutOAuthRegistration(ctx context.Context, req *proto.OAuthRegistration) (*emptypb.Empty, error) {
	store, ok := s.store.(OAuthRegistrationStore)
	if !ok {
		return nil, providerRPCError("put oauth registration", ErrOAuthRegistrationStoreUnsupported)
	}
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	if err := store.PutOAuthRegistration(ctx, req); err != nil {
		return nil, providerRPCError("put oauth registration", err)
	}
	return &emptypb.Empty{}, nil
}

func (s *datastoreServer) DeleteOAuthRegistration(ctx context.Context, req *proto.DeleteOAuthRegistrationRequest) (*emptypb.Empty, error) {
	store, ok := s.store.(OAuthRegistrationStore)
	if !ok {
		return nil, providerRPCError("delete oauth registration", ErrOAuthRegistrationStoreUnsupported)
	}
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	if err := store.DeleteOAuthRegistration(ctx, req.GetAuthServerUrl(), req.GetRedirectUri()); err != nil {
		return nil, providerRPCError("delete oauth registration", err)
	}
	return &emptypb.Empty{}, nil
}

File: sdk/rust/src/datastore_server.rs
use std::sync::Arc;

use tonic::{Request as GrpcRequest, Response as GrpcResponse, Status};

use crate::datastore::DatastoreProvider;
use crate::generated::v1::datastore_provider_server::DatastoreProvider as DatastoreProviderGrpc;
use crate::generated::v1::{
    DeleteOAuthRegistrationRequest, DeleteStoredIntegrationTokenRequest, GetApiTokenByHashRequest,
    GetOAuthRegistrationRequest, GetStoredIntegrationTokenRequest, GetUserRequest,
    ListApiTokensRequest, ListApiTokensResponse, ListStoredIntegrationTokensRequest,
    ListStoredIntegrationTokensResponse, OAuthRegistration, RevokeAllApiTokensRequest,
    RevokeAllApiTokensResponse, RevokeApiTokenRequest, StoredApiToken, StoredIntegrationToken,
    StoredUser,
};
use crate::rpc_status::rpc_status;

#[derive(Clone)]
pub struct DatastoreServer<P> {
    store: Arc<P>,
}

impl<P> DatastoreServer<P> {
    pub fn new(store: Arc<P>) -> Self {
        Self { store }
    }
}

#[tonic::async_trait]
impl<P> DatastoreProviderGrpc for DatastoreServer<P>
where
    P: DatastoreProvider,
{
    async fn migrate(
        &self,
        _request: GrpcRequest<()>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        self.store
            .migrate()
            .await
            .map_err(|error| rpc_status("migrate", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn get_user(
        &self,
        request: GrpcRequest<GetUserRequest>,
    ) -> std::result::Result<GrpcResponse<StoredUser>, Status> {
        let request = request.into_inner();
        let user = self
            .store
            .get_user(&request.id)
            .await
            .map_err(|error| rpc_status("get user", error))?;
        let Some(user) = user else {
            return Err(Status::not_found("user not found"));
        };
        Ok(GrpcResponse::new(user))
    }

    async fn find_or_create_user(
        &self,
        request: GrpcRequest<crate::generated::v1::FindOrCreateUserRequest>,
    ) -> std::result::Result<GrpcResponse<StoredUser>, Status> {
        let request = request.into_inner();
        let user = self
            .store
            .find_or_create_user(&request.email)
            .await
            .map_err(|error| rpc_status("find or create user", error))?;
        Ok(GrpcResponse::new(user))
    }

    async fn put_stored_integration_token(
        &self,
        request: GrpcRequest<StoredIntegrationToken>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        self.store
            .put_integration_token(request.into_inner())
            .await
            .map_err(|error| rpc_status("put integration token", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn get_stored_integration_token(
        &self,
        request: GrpcRequest<GetStoredIntegrationTokenRequest>,
    ) -> std::result::Result<GrpcResponse<StoredIntegrationToken>, Status> {
        let request = request.into_inner();
        let token = self
            .store
            .get_integration_token(
                &request.user_id,
                &request.integration,
                &request.connection,
                &request.instance,
            )
            .await
            .map_err(|error| rpc_status("get integration token", error))?;
        let Some(token) = token else {
            return Err(Status::not_found("integration token not found"));
        };
        Ok(GrpcResponse::new(token))
    }

    async fn list_stored_integration_tokens(
        &self,
        request: GrpcRequest<ListStoredIntegrationTokensRequest>,
    ) -> std::result::Result<GrpcResponse<ListStoredIntegrationTokensResponse>, Status> {
        let request = request.into_inner();
        let tokens = self
            .store
            .list_integration_tokens(&request.user_id, &request.integration, &request.connection)
            .await
            .map_err(|error| rpc_status("list integration tokens", error))?;
        Ok(GrpcResponse::new(ListStoredIntegrationTokensResponse {
            tokens,
        }))
    }

    async fn delete_stored_integration_token(
        &self,
        request: GrpcRequest<DeleteStoredIntegrationTokenRequest>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        let request = request.into_inner();
        self.store
            .delete_integration_token(&request.id)
            .await
            .map_err(|error| rpc_status("delete integration token", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn put_api_token(
        &self,
        request: GrpcRequest<StoredApiToken>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        self.store
            .put_api_token(request.into_inner())
            .await
            .map_err(|error| rpc_status("put api token", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn get_api_token_by_hash(
        &self,
        request: GrpcRequest<GetApiTokenByHashRequest>,
    ) -> std::result::Result<GrpcResponse<StoredApiToken>, Status> {
        let request = request.into_inner();
        let token = self
            .store
            .get_api_token_by_hash(&request.hashed_token)
            .await
            .map_err(|error| rpc_status("get api token by hash", error))?;
        let Some(token) = token else {
            return Err(Status::not_found("api token not found"));
        };
        Ok(GrpcResponse::new(token))
    }

    async fn list_api_tokens(
        &self,
        request: GrpcRequest<ListApiTokensRequest>,
    ) -> std::result::Result<GrpcResponse<ListApiTokensResponse>, Status> {
        let request = request.into_inner();
        let tokens = self
            .store
            .list_api_tokens(&request.user_id)
            .await
            .map_err(|error| rpc_status("list api tokens", error))?;
        Ok(GrpcResponse::new(ListApiTokensResponse { tokens }))
    }

    async fn revoke_api_token(
        &self,
        request: GrpcRequest<RevokeApiTokenRequest>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        let request = request.into_inner();
        self.store
            .revoke_api_token(&request.user_id, &request.id)
            .await
            .map_err(|error| rpc_status("revoke api token", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn revoke_all_api_tokens(
        &self,
        request: GrpcRequest<RevokeAllApiTokensRequest>,
    ) -> std::result::Result<GrpcResponse<RevokeAllApiTokensResponse>, Status> {
        let request = request.into_inner();
        let revoked = self
            .store
            .revoke_all_api_tokens(&request.user_id)
            .await
            .map_err(|error| rpc_status("revoke all api tokens", error))?;
        Ok(GrpcResponse::new(RevokeAllApiTokensResponse { revoked }))
    }

    async fn get_o_auth_registration(
        &self,
        request: GrpcRequest<GetOAuthRegistrationRequest>,
    ) -> std::result::Result<GrpcResponse<OAuthRegistration>, Status> {
        let request = request.into_inner();
        let registration = self
            .store
            .get_oauth_registration(&request.auth_server_url, &request.redirect_uri)
            .await
            .map_err(|error| rpc_status("get oauth registration", error))?;
        let Some(registration) = registration else {
            return Err(Status::not_found("oauth registration not found"));
        };
        Ok(GrpcResponse::new(registration))
    }

    async fn put_o_auth_registration(
        &self,
        request: GrpcRequest<OAuthRegistration>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        self.store
            .put_oauth_registration(request.into_inner())
            .await
            .map_err(|error| rpc_status("put oauth registration", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn delete_o_auth_registration(
        &self,
        request: GrpcRequest<DeleteOAuthRegistrationRequest>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        let request = request.into_inner();
        self.store
            .delete_oauth_registration(&request.auth_server_url, &request.redirect_uri)
            .await
            .map_err(|error| rpc_status("delete oauth registration", error))?;
        Ok(GrpcResponse::new(()))
    }
}
